tests/soft_limiter_tests.cpp

```diff
@@ -55,7 +55,11 @@ TEST(SoftLimiter, InboundsProcessTooManyFrames)
 	SoftLimiter limiter("test-channel");
 	const std::vector<float> in{-3, -2, -1, 0, 1, 2};
 	std::vector<int16_t> out(frames * 2);
+#ifdef _GLIBCXX_ASSERTIONS
+	EXPECT_DEATH({ limiter.Process(in, frames + 1, out); }, "");
+#else
 	EXPECT_DEBUG_DEATH({ limiter.Process(in, frames + 1, out); }, "");
+#endif
```

Comment:
We expect death in every case:
 - If this is a build with all asserts off, then going beyond the vector should segfault/bus-error.
 - If this is a debug build, then the soft-limiter's bound assert(s) will terminate
 - If this is a non-debug build with a checked STL, then the vector's bound assert(s) will terminate

Therefore, we can have a single call to: EXPECT_DEATH, without the need for pre-processor branches.

Reply:
> We expect death in every case:
>     * If this is a build with all asserts off, then going beyond the vector should segfault/bus-error.
>     * If this is a debug build, then the soft-limiter's bound assert(s) will terminate
>     * If this is a non-debug build with a checked STL, then the vector's bound assert(s) will terminate
> 
> Therefore, we can have a single call to `EXPECT_DEATH` without pre-processor branches.

Unfortunately, the first point does not hold. There are quite many things you can do with a `std::vector` that respond to out-of-bound with undefined behavior, which does not necessarily (or even usually) mean crashing — more likely adjacent memory is accessed like the vector extended there. Segfault only happens if said memory location is not allocated for the process at all.

So if you change this assertion to `EXPECT_DEATH`, compile with `buildtype=plain` and run the test, it fails and the test log contains this:

```
[ RUN      ] SoftLimiter.InboundsProcessTooManyFrames
../tests/soft_limiter_tests.cpp:58: Failure
Death test: { limiter.Process(in, frames + 1, out); }
    Result: failed to die.
 Error msg:
[  DEATH   ] 
[  FAILED  ] SoftLimiter.InboundsProcessTooManyFrames (0 ms)
```

Because it is undefined behavior, it depends on the compiler. I have GCC 11.1.1.

This is also exactly what `GLIBCXX_ASSERTIONS` changes. It sets the undefined behavior of certain operations to mean a crash.

If you really want no difference of behavior between debug and release builds, and it is ok to crash the program, `assert(condition)` statements could be replaced with this kind of code: `if (!condition) { abort() }`. Then there would be no undefined behavior. Calling `abort()` could be replaced with a graceful recovery action, but writing that action would need understanding of the soft limiter. That is beyond my abiility at the moment, but if you are content with crashing, I can rewrite the asserts into the form proposed above. What do you think?

Reply:
We prefer assert checks as internal "checks-and-balances" when we know the condition is internally-controlled code/logic.  However when potentially bad data or conditions are outside of our control (system or user-data), we prefer non-asserts, such as a conditional+abort. 

This lets us be very liberal with asserting expected states, which also acts as self-describing code to readers. This is much more powerful than a comment like "// We expect the values to be positive at this point", while gaining the performance in release builds when those asserts disappear.

In this case, I was trying to avoid proliferation of pre-processor branches in our unit tests (perhaps Clang, VS2019, musl libc, etc.. have their own variations on this ifdef as well.. which would make our unit tests quite a mess).   

Given Fedora is using this flag by default, and Fedora is 1st-class certainly a supported system, we want to make sure all Fedora users can run unit tests - regardless of build-type.

@LowLevelMahn - I like your suggestion to abstract this behind a macro.  Right now, this is our only instance - so when the day comes that more of this is needed, then you're suggestion will be put into action :-)

Merging :rocket: 

Thank you,  @oturpe!

Reply:
The test passes on Ubuntu using all build-types:

![2021-06-20_19-06](https://user-images.githubusercontent.com/1557255/122698480-3cfb5480-d1fc-11eb-8ff2-8d3880d8aa33.png)
![2021-06-20_19-02](https://user-images.githubusercontent.com/1557255/122698481-3d93eb00-d1fc-11eb-9ded-4af9a8352277.png)